composer/trainer/trainer.py

```diff
@@ -2193,3 +2194,39 @@ def save_checkpoint(self, name: str = 'ep{epoch}-ba{batch}-rank{rank}', *, weigh
             List[pathlib.Path]: See :func:`.save_checkpoint`.
         """
         return save_checkpoint(state=self.state, filename=name, weights_only=weights_only)
+
+    def export_for_inference(
+        self,
+        save_format: Union[str, ExportFormat],
+        save_path: str,
+        save_object_store: Optional[ObjectStore] = None,
+        sample_input: Optional[Any] = None,
+        transforms: Optional[Sequence[Transform]] = None,
+    ):
+        """Export a model for inference.
+
+        Args:
+            save_format (Union[str, ExportFormat]):  Format to export to. Either ``"torchscript"`` or ``"onnx"``.
+            save_path: (str): The path for storing the exported model. It can be a path to a file on the local disk,
+            a URL, or if ``save_object_store`` is set, the object name
+                in a cloud bucket. For example, ``my_run/exported_model``.
+            save_object_store (ObjectStore, optional): If the ``save_path`` is in an object name in a cloud bucket
+                (i.e. AWS S3 or Google Cloud Storage), an instance of
+                :class:`~.ObjectStore` which will be used
+                to store the exported model. Set this to ``None`` if ``save_path`` is a local filepath.
+                (default: ``None``)
+            sample_input (Any, optional): Example model inputs used for tracing. This is needed for "onnx" export.
+                The ``sample_input`` need not match the batch size you intend to use for inference. However, the model
+                should accept the ``sample_input`` as is. (default: ``None``)
+            transforms (Sequence[Transform], optional): transformations (usually optimizations) that should
+                be applied to the model. Each Transform should be a callable that takes a model and returns a modified model.
+
+        Returns:
+            None
+        """
+        inference.export_for_inference(model=self.state.model,
+                                       save_format=save_format,
+                                       save_path=save_path,
+                                       save_object_store=save_object_store,
```

Comment:
Also if save_object_store is not given, it should default to the one given to trainer. See 'CheckpointSaver' callback on how it stores checkpoint.

Reply:
If we do this, it could be helpful & convenient to users. However, it will break the pattern of the other ExportForInference functions ([composer.callbacks.ExportForInferenceCallBack](https://github.com/mosaicml/composer/blob/4f00c2205c800b4962bfeb110153ee566db79e15/composer/callbacks/export_for_inference.py#L49) and [composer.utils.export_for_inference](https://github.com/mosaicml/composer/blob/4f00c2205c800b4962bfeb110153ee566db79e15/composer/utils/inference.py#L68)), for which save_object_store is either an ObjectStore object (if the user is saving remotely), or None, if the user wishes to save locally.
Implementing it this way would result in different behavior for
```
trainer = Trainer(
    model=MNIST_Classifier(num_classes=10),
		train_dataloader=train_dataloader,
		callbacks=ExportForInferenceCallback(save_path='model/",save_format='torchscript')
		load_path='/tmp/ep1-rank0',
		loggers='ObjectStoreLogger('s3',..)'
)
```
and
```
trainer = Trainer(
    model=MNIST_Classifier(num_classes=10),
		train_dataloader=train_dataloader,
		load_path='/tmp/ep1-rank0',
		loggers='ObjectStoreLogger('s3',...)'
)
trainer.export_for_inference(save_path='model/",save_format='torchscript')
```
One would be saved locally and one would not.

Reply:
@ravi-mosaicml or @hanlint. Any thoughts?

Reply:
Discussed offline but leaving my thoughts here too: IMO this discrepancy is ok since` trainer.export_for_infernce ` method is tied to trainer and user is most likely to use the same object store for both.

 
If the trainer is saving checkpoints locally, `trainer.export_for_inference` will also export locally

if the trainer is saving checkpoints in a remote store, `trainer.export_for_inference` will also export to the same remote store

Reply:
Looking at the code above, I can't actually tell which example would save locally and which one would not, which seems to me we should maintain consistent behavior across both. In both cases, I would expect the method to only save locally, and never export? 

Better to require explicit, rather than be convenient but unclear?

Reply:
@nik-mosaic In your examples above, since loggers is S3 object store both would store using this logger. Isn't it?

Reply:
Because save_object_store is not passed into the ExportForInferenceCallback, it will be none, which will call composer.utils.export_for_inference with save_object_store=None, which will then cause it to be stored locally, since we have `is_remote_store = save_object_store is not None`.

Reply:
Oh! I thought callback also uses loggers. I think we should use loggers for `ExportForInferenceCallback` too.

Reply:
By doing so we will make exporting functionality similar to checkpoint_saver functionality that also uses logger for checkpoint saving.

Reply:
I can definitely see the benefit of making the decision to use loggers across the board, but that's probably best done in a separate PR to keep the scope of this one limited to the trainer Export wrapper.

Reply:
Sure. feel free to do it in a separate PR. Could you please create a JIRA so that we don't forget? I was hoping we could add it all for the release today along with notebook update https://github.com/mosaicml/composer/pull/1360

Reply:
I don't think we should need to take an object store here; instead, this function can call `logger.file_artifact(save_path)`. That will then use any/all artifact loggers specified as part of the trainer.

Reply:
This is just for overriding, in case, a user want to export model to a specific object store separate from training checkpoints. if None, it will use `logger.file_artifact`.